File: Data Structures & Algorithms/subsets/submission-6.cpp
class Solution {
public:
    vector<vector<int>> res;
    vector<int> subset;
    vector<vector<int>> subsets(vector<int>& nums) {
        //[1] or []

        dfs(0, res, nums);
        return res;
    }

void dfs(int index, vector<vector<int>>& res, vector<int>& nums) {
    res.push_back(subset);
    for (int i=index; i<nums.size(); i++) {
        subset.push_back(nums[i]);
        dfs(i+1, res, nums);
        subset.pop_back();
        }
    }
};

File: Data Structures & Algorithms/combination-target-sum-ii/submission-6.py
class Solution:
    def combinationSum2(self, candidates: List[int], target: int) -> List[List[int]]:
        res, subset = [], []
        candidates.sort()
        sum = 0

        def dfs(index, sum):
            if sum>target: return
            if sum == target:
                res.append(subset.copy())
                return
            
            for i in range(index, len(candidates)):
                if (i>index and candidates[i] == candidates[i-1]): continue
                subset.append(candidates[i])
                dfs(i+1, sum+candidates[i])
                subset.pop()

        dfs(0, 0)
        return res